File: Algorithms/p199_Binary_Tree_Right_Side_View_2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        self.rightSideViewHelper(root,res,0)
        return res
        
    def rightSideViewHelper(self, root, res, level):
        if not root:
            return
        if level==len(res):
            res.append(root.val)
        self.rightSideViewHelper(root.right,res,level+1)
        self.rightSideViewHelper(root.left,res,level+1)

File: Algorithms/p199_Binary_Tree_Right_Side_View_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        rightSideViewHelper(root,res,0);
        return res;
    }
    void rightSideViewHelper(TreeNode* root, vector<int>& res, int level){
        if(!root) return;
        if(level==res.size()) res.push_back(root->val);
        rightSideViewHelper(root->right,res,level+1);
        rightSideViewHelper(root->left,res,level+1);
    }
};
